xds/src/main/java/io/grpc/xds/internal/certprovider/MeshCaCertificateProvider.java

```diff
@@ -17,35 +17,349 @@
+      logger.log(Level.SEVERE, "Inside sleep", ex);
+    }
+  }
+
+  private static boolean retriable(Throwable t) {
+    if (t instanceof StatusRuntimeException) {
+      return retriable(((StatusRuntimeException) t).getStatus());
+    } else if (t.getCause() instanceof StatusRuntimeException) {
+      return retriable(((StatusRuntimeException) t.getCause()).getStatus());
+    }
+    return false;
+  }
+
+  private static boolean retriable(Status status) {
+    return RETRIABLE_CODES.contains(status.getCode());
+  }
+
+  private void generateErrorIfCurrentCertExpired(Throwable t) {
+    X509Certificate currentCert = getWatcher().getLastIdentityCert();
+    if (currentCert != null) {
+      long delaySeconds = computeDelaySecondsToCertExpiry(currentCert);
+      if (delaySeconds > INITIAL_DELAY_SECONDS) {
+        return;
+      }
+      getWatcher().cleanupLastValues();
+    }
+    getWatcher().onError(Status.fromThrowable(t));
+  }
+
+  private KeyPair generateKeyPair() throws NoSuchAlgorithmException {
+    KeyPairGenerator keyPairGenerator = KeyPairGenerator.getInstance(alg);
+    keyPairGenerator.initialize(keySize);
+    return keyPairGenerator.generateKeyPair();
+  }
+
+  private String generateCsr(KeyPair pair) throws IOException, OperatorCreationException {
+    PKCS10CertificationRequestBuilder p10Builder =
+        new JcaPKCS10CertificationRequestBuilder(
+            new X500Principal("CN=EXAMPLE.COM"), pair.getPublic());
+    JcaContentSignerBuilder csBuilder = new JcaContentSignerBuilder(signatureAlg);
+    ContentSigner signer = csBuilder.build(pair.getPrivate());
+    PKCS10CertificationRequest csr = p10Builder.build(signer);
+    PemObject pemObject = new PemObject("NEW CERTIFICATE REQUEST", csr.getEncoded());
+    StringWriter str = new StringWriter();
+    JcaPEMWriter pemWriter = new JcaPEMWriter(str);
+    pemWriter.writeObject(pemObject);
+    pemWriter.close();
+    str.close();
+    return str.toString();
+  }
+
+  /** Compute refresh interval as half of interval to current cert expiry. */
+  private long computeRefreshSecondsFromCurrentCertExpiry() {
+    X509Certificate lastCert = getWatcher().getLastIdentityCert();
+    if (lastCert == null) {
+      return INITIAL_DELAY_SECONDS;
+    }
+    long delayToCertExpirySeconds = computeDelaySecondsToCertExpiry(lastCert) / 2;
+    return Math.max(delayToCertExpirySeconds, INITIAL_DELAY_SECONDS);
+  }
+
+  private long computeDelaySecondsToCertExpiry(X509Certificate lastCert) {
+    checkNotNull(lastCert, "lastCert");
+    return TimeUnit.NANOSECONDS.toSeconds(
+        TimeUnit.MILLISECONDS.toNanos(lastCert.getNotAfter().getTime()) - timeProvider
+            .currentTimeNanos());
+  }
+
+  private static void shutdownChannel(ManagedChannel channel) {
+    channel.shutdown();
+    try {
+      channel.awaitTermination(10, TimeUnit.SECONDS);
+    } catch (InterruptedException ex) {
+      logger.log(Level.SEVERE, "awaiting channel Termination", ex);
```

Comment:
Always call `Thread.currentThread().interrupt()` to re-mark the thread as interrupted. Ditto elsewhere.

Reply:
Good point. Done